### PR TITLE
Update FiniteDifferences

### LINKED ISSUE #18: Update FiniteDifferences
We've currently got this package and, by extension, ChainRules.jl fixed to FiniteDifferences v0.7. This is bad because there are newer and better versions available. There was a breaking API change between 0.7 and 0.9, so would be good to upgrade and change bounds.

### PR DESCRIPTION
**Pull request by nickrobinson251**
- closes #18 
- requires https://github.com/JuliaDiff/ChainRulesCore.jl/pull/135

### CODE REVIEW

File: src/ChainRulesTestUtils.jl [hunk 1]
```diff
@@ -11,8 +11,12 @@ const _fdm = central_fdm(5, 1)
 
 export test_scalar, frule_test, rrule_test, isapprox, generate_well_conditioned_matrix
 
+# TODO: recondiser these https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/7
```
Comment (mattBrzezinski) [inline src/ChainRulesTestUtils.jl]:
```suggestion
# TODO: reconsider these https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/7
```

File: test/runtests.jl [hunk 2]
```diff
@@ -4,20 +4,48 @@ using Random
...
+    @testset "unary: identity(x)" begin
+        function ChainRulesCore.frule((_, ẏ), ::typeof(identity), x)
+            return x, ẏ
+        end
+        function ChainRulesCore.rrule(::typeof(identity), x)
+            function identity_pullback(ȳ)
+                return (NO_FIELDS, ȳ)
+            end
+            return x, identity_pullback
+        end
```
Comment (mattBrzezinski) [inline test/runtests.jl]:
```suggestion
    @testset "unary: identity(x)" begin
        function ChainRulesCore.frule((_, ẏ), ::typeof(identity), x)
            return x, ẏ
        end
        
        function ChainRulesCore.rrule(::typeof(identity), x)
            function identity_pullback(ȳ)
                return (NO_FIELDS, ȳ)
            end
            
            return x, identity_pullback
        end
```

File: test/runtests.jl [hunk 3]
```diff
@@ -4,20 +4,48 @@ using Random
...
+        @testset "frule_test" begin
+            frule_test(fst, (2, 4.0), (3, 5.0))
+            frule_test(fst, (randn(4), randn(4)), (randn(4), randn(4)))
+        end
+        @testset "rrule_test" begin
+            rrule_test(fst, rand(), (2.0, 4.0), (3.0, 5.0))
+            rrule_test(fst, randn(4), (randn(4), randn(4)), (randn(4), randn(4)))
+        end
+    end
```
Comment (mattBrzezinski) [inline test/runtests.jl]:
```suggestion
    @testset "binary: fst(x, y)" begin
        fst(x, y) = x
        ChainRulesCore.frule((_, dx, dy), ::typeof(fst), x, y) = (x, dx)
        
        function ChainRulesCore.rrule(::typeof(fst), x, y)
            function fst_pullback(Δx)
                return (NO_FIELDS, Δx, Zero())
            end
        
            return x, fst_pullback
        end
        
        @testset "frule_test" begin
            frule_test(fst, (2, 4.0), (3, 5.0))
            frule_test(fst, (randn(4), randn(4)), (randn(4), randn(4)))
        end
        
        @testset "rrule_test" begin
            rrule_test(fst, rand(), (2.0, 4.0), (3.0, 5.0))
            rrule_test(fst, randn(4), (randn(4), randn(4)), (randn(4), randn(4)))
        end
    end
```

**Comment by mattBrzezinski** [comment]
The failure from TravisCI:

```
ERROR: Unsatisfiable requirements detected for package ChainRulesCore [d360d2e6]:
 ChainRulesCore [d360d2e6] log:
 ├─possible versions are: [0.1.0-0.1.1, 0.2.0, 0.3.0, 0.4.0, 0.5.0-0.5.5, 0.6.0-0.6.1, 0.7.0] or uninstalled
 └─restricted to versions 0.7.1-0.7 by an explicit requirement — no versions left
```

**Comment by nickrobinson251** [comment]
Yep that's expected, as this first requires JuliaDiff/ChainRulesCore.jl#135

**Pull request merged**